### PR TITLE
Handle TransferToEthereum

### PR DESCRIPTION
**Pull request by yito88**
- Add the inserted height to the Merkle tree of the bridge pool for the timeout checking
- Handle TransferToEthereum event
  - Delete the pending transfers contained in the event
  - Refund and delete for timed-out transfers

### CODE REVIEW

File: ethereum_bridge/src/protocol/transactions/ethereum_events/events.rs [hunk 1]
```diff
@@ -84,22 +103,263 @@ where
...
+    for event in transfers {
+        let pending_transfer = event.into();
+        let key = get_pending_key(&pending_transfer);
+        if storage.has_key(&key)?.0 {
+            _ = storage.delete(&key)?;
+            _ = pending_keys.remove(&key);
+        } else {
+            return Err(eyre::eyre!(
+                "The transfer doesn't exist in the bridge pool"
+            ));
```
Comment (sug0) [inline ethereum_bridge/src/protocol/transactions/ethereum_events/events.rs:136]: I think the only way we could have run into this code path is if a quorum of validators decided on an event that didn't happen (that is, effectively, we would have had a Byzantine quorum of validators at that point). to transfer to Ethereum, users need to provide a proof that they sent a transfer to the bridge pool, at which point their transfer would be registered in storage.

therefore, maybe it's wiser to panic here? or use `unreachable!()`? there's also `namada_core::hints::likely()`, for the optimistic if branch.

**Review by sug0 — COMMENTED** [review]
a few comments, but otherwise lgtm

File: core/src/ledger/eth_bridge/storage/bridge_pool.rs [hunk 2]
```diff
@@ -95,16 +105,31 @@ impl BridgePoolTree {
...
+            eyre!("Could not parse key segment as a hash").into()
+        })
     }
 
     /// Update the tree with a new value.
     ///
     /// Returns the new root if successful. Will
     /// return an error if the key is malformed.
-    pub fn insert_key(&mut self, key: &Key) -> Result<Hash, Error> {
+    pub fn insert_key(
```
Comment (batconjurer) [inline core/src/ledger/eth_bridge/storage/bridge_pool.rs:126]: nit: The rest of the trees follow a key-value order for the insert arguments.

File: core/src/ledger/storage/merkle_tree.rs [hunk 3]
```diff
@@ -324,12 +322,15 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
     fn update_tree(
```
Comment (batconjurer) [inline core/src/ledger/storage/merkle_tree.rs:322]: Let s just the make the height the value of the tree.

File: core/src/ledger/storage/traits.rs [hunk 4]
```diff
@@ -60,6 +63,13 @@ impl<'a, H: StorageHasher + Default> SubTreeRead for &'a Smt<H> {
         }
     }
 
+    fn subtree_inserted_height(
+        &self,
+        _key: &Key,
+    ) -> Result<BlockHeight, Error> {
+        Err(Error::MerkleTree("SMT can't store the height".to_string()))
+    }
+
```
[the remark this answers is not in the document]
Comment (batconjurer) [reply core/src/ledger/storage/traits.rs:72]: I think we should just tree the block height as the value type of the tree. Then we can use all the existing interfaces.

**Review by batconjurer — CHANGES_REQUESTED** [review]

File: core/src/ledger/storage/merkle_tree.rs [hunk 5]
```diff
@@ -344,10 +342,28 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
...
 
+    /// Get the value in the tree
+    pub fn get(&self, key: &Key) -> Result<Vec<u8>> {
+        let (store_type, sub_key) = StoreType::sub_key(key)?;
+        self.tree(&store_type).subtree_get(&sub_key)
+    }
+
     /// Update the tree with the given key and value
-    pub fn update(&mut self, key: &Key, value: impl AsRef<[u8]>) -> Result<()> {
+    pub fn update(
```
Comment (batconjurer) [inline core/src/ledger/storage/merkle_tree.rs:352]: can't we just pass the blockheight in as the value parameter here?

**Review by batconjurer — CHANGES_REQUESTED** [review]

File: ethereum_bridge/src/protocol/transactions/ethereum_events/events.rs [hunk 6]
```diff
@@ -84,22 +104,263 @@ where
...
+    // Remove the completed transfers from the bridge pool
+    for event in transfers {
+        let pending_transfer = event.into();
+        let key = get_pending_key(&pending_transfer);
+        if likely(storage.has_key(&key)?.0) {
+            _ = storage.delete(&key)?;
+            _ = pending_keys.remove(&key);
+        } else {
+            // The transfer should exist in the bridge pool
+            unreachable!()
```
Comment (sug0) [inline ethereum_bridge/src/protocol/transactions/ethereum_events/events.rs:136]: nit: you can add the message in the comment to the panic itself
```suggestion
            unreachable!("The transfer should exist in the bridge pool");
```

**Review by sug0 — APPROVED** [review]
lgtm, just a small suggestion

**Review by sug0 — APPROVED** [review]

**Review by batconjurer — APPROVED** [review]

**Review by batconjurer — CHANGES_REQUESTED** [review]
I have thought of a mild security concern that should be addressed actually. WIll leave comments in the relevant places.

**Review by batconjurer — APPROVED** [review]
I have thought of a mild security concern that should be addressed actually. WIll leave comments in the relevant places.

**Comment by batconjurer** [comment]
> I have thought of a mild security concern that should be addressed actually. WIll leave comments in the relevant places.

Actually, it's fine.

**Pull request merged**